python/beeai_framework/context.py

```diff
@@ -49,27 +51,31 @@ class Run(Generic[R]):
     def __init__(self, handler: Callable[[], R | Awaitable[R]], context: "RunContext") -> None:
         super().__init__()
         self.handler = ensure_async(handler)
-        self.tasks: list[tuple[Callable[[Any], None], Any]] = []
+        self.tasks: list[tuple[Callable, list]] = []
         self.run_context = context
 
     def __await__(self) -> Generator[Any, None, R]:
         return self._run_tasks().__await__()
 
-    def observe(self, fn: Callable[[Emitter], Any]) -> Self:
-        self.tasks.append((fn, self.run_context.emitter))
+    def observe(self, fn: Callable[[Emitter], None]) -> Self:
+        self.tasks.append((fn, [self.run_context.emitter]))
+        return self
+
+    def on(self, matcher: Matcher, callback: Callback, options: EmitterOptions | None = None) -> Self:
+        self.tasks.append((self.run_context.emitter.match, [matcher, callback, options]))
         return self
 
     def context(self, context: dict) -> Self:
-        self.tasks.append((self._set_context, context))
+        self.tasks.append((self._set_context, [context]))
         return self
 
     def middleware(self, fn: Callable[["RunContext"], None]) -> Self:
-        self.tasks.append((fn, self.run_context))
+        self.tasks.append((fn, [self.run_context]))
         return self
 
     async def _run_tasks(self) -> R:
-        for fn, param in self.tasks:
-            await ensure_async(fn)(param)
+        for fn, params in self.tasks:
+            await ensure_async(fn)(*params)
 
         self.tasks.clear()
```

Comment:
clear should be done in advance

Reply:
What do you mean? clear() empties the list and we iterate over the list in the chunk above this. Where are we supposed the do it?

Reply:
```
tasks = self.tasks.splice()
for fn, params in tasks:
  await ensure_async(fn)(*params)
```

Reply:
but it should be ok as it is now

Reply:
I think I understand, since the callbacks can be asynchronous you want to copy the task list, clear self.list, then run the tasks from the copy:
```python
    async def _run_tasks(self) -> R:
        tasks = self.tasks[:]
        self.tasks.clear()

        for fn, params in tasks:
            await ensure_async(fn)(*params)

        return await self.handler()
```

Reply:
done